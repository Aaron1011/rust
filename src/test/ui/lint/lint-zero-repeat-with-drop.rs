#![feature(const_generics)]
#![allow(incomplete_features)]
#![deny(zero_repeat_with_drop)]

const ZERO: usize = 1 * 0;

const fn make_val<T>(val: T) -> T { val }

struct NoDropOrCopy;
struct WithDropGlue(String);

fn foo<T, V: Copy, F: Fn() -> T, const N: usize>(not_copy: F, copy: V) {
    // All of these should triger the lint
    let _ = [not_copy(); 0]; //~ ERROR used a type
    let _ = [not_copy(); 1 - 1]; //~ ERROR used a type
    let _ = [not_copy(); ZERO]; //~ ERROR used a type
    let _ = [Some(not_copy()); 0]; //~ ERROR used a type
    let _ = [None::<T>; 0]; //~ ERROR used a type
    let _ = [make_val(not_copy()); 0]; //~ ERROR used a type
    let _ = [String::new(); 0]; //~ ERROR used a type
    let _ = [WithDropGlue(String::new()); 0]; //~ ERROR used a type

    // None of these should trigger the lint
    let _ = [copy; 0];
    let _ = [Some(copy); 0];
    let _ = [NoDropOrCopy; 0];
    let _ = [not_copy(); 1];
    let _ = [copy; N];
}

fn allow_it() {
    #[allow(zero_repeat_with_drop)]
    let _ = [WithDropGlue(String::new()); 1 - 1];
}

fn main() {}
