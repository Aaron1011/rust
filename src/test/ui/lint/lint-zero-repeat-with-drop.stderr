error: used a type with a destructor in a zero-length repeat expression
  --> $DIR/lint-zero-repeat-with-drop.rs:14:13
   |
LL |     let _ = [not_copy(); 0];
   |             ^^^^^^^^^^^^^^^ help: consider using an empty array expression: `[]`
   |
note: the lint level is defined here
  --> $DIR/lint-zero-repeat-with-drop.rs:3:9
   |
LL | #![deny(zero_repeat_with_drop)]
   |         ^^^^^^^^^^^^^^^^^^^^^
   = note: the value used here has type `T`, which may have a destructor
   = note: a length of zero is used, which will cause the value to be dropped in a strange location

error: used a type with a destructor in a zero-length repeat expression
  --> $DIR/lint-zero-repeat-with-drop.rs:15:13
   |
LL |     let _ = [not_copy(); 1 - 1];
   |             ^^^^^^^^^^^^^^^^^^^ help: consider using an empty array expression: `[]`
   |
   = note: the value used here has type `T`, which may have a destructor
   = note: a length of zero is used, which will cause the value to be dropped in a strange location

error: used a type with a destructor in a zero-length repeat expression
  --> $DIR/lint-zero-repeat-with-drop.rs:16:13
   |
LL |     let _ = [not_copy(); ZERO];
   |             ^^^^^^^^^^^^^^^^^^ help: consider using an empty array expression: `[]`
   |
   = note: the value used here has type `T`, which may have a destructor
   = note: a length of zero is used, which will cause the value to be dropped in a strange location

error: used a type with a destructor in a zero-length repeat expression
  --> $DIR/lint-zero-repeat-with-drop.rs:17:13
   |
LL |     let _ = [Some(not_copy()); 0];
   |             ^^^^^^^^^^^^^^^^^^^^^ help: consider using an empty array expression: `[]`
   |
   = note: the value used here has type `std::option::Option<T>`, which may have a destructor
   = note: a length of zero is used, which will cause the value to be dropped in a strange location

error: used a type with a destructor in a zero-length repeat expression
  --> $DIR/lint-zero-repeat-with-drop.rs:18:13
   |
LL |     let _ = [None::<T>; 0];
   |             ^^^^^^^^^^^^^^ help: consider using an empty array expression: `[]`
   |
   = note: the value used here has type `std::option::Option<T>`, which may have a destructor
   = note: a length of zero is used, which will cause the value to be dropped in a strange location

error: used a type with a destructor in a zero-length repeat expression
  --> $DIR/lint-zero-repeat-with-drop.rs:19:13
   |
LL |     let _ = [make_val(not_copy()); 0];
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider using an empty array expression: `[]`
   |
   = note: the value used here has type `T`, which may have a destructor
   = note: a length of zero is used, which will cause the value to be dropped in a strange location

error: used a type with a destructor in a zero-length repeat expression
  --> $DIR/lint-zero-repeat-with-drop.rs:20:13
   |
LL |     let _ = [String::new(); 0];
   |             ^^^^^^^^^^^^^^^^^^ help: consider using an empty array expression: `[]`
   |
   = note: the value used here has type `std::string::String`, which may have a destructor
   = note: a length of zero is used, which will cause the value to be dropped in a strange location

error: used a type with a destructor in a zero-length repeat expression
  --> $DIR/lint-zero-repeat-with-drop.rs:21:13
   |
LL |     let _ = [WithDropGlue(String::new()); 0];
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider using an empty array expression: `[]`
   |
   = note: the value used here has type `WithDropGlue`, which may have a destructor
   = note: a length of zero is used, which will cause the value to be dropped in a strange location

error: aborting due to 8 previous errors

